homeassistant/components/media_player/yamaha.py

```diff
@@ -49,8 +57,40 @@ def update(self):
             self._pwstate = STATE_OFF
         self._muted = self._receiver.mute
         self._volume = (self._receiver.volume/100) + 1
-        self._current_source = self._receiver.input
-        self._source_list = list(self._receiver.inputs().keys())
+
+        if self.source_list is None:
+            self.build_source_list()
+
+        current_source = self._receiver.input
+        if self._mapping is not None:
+            if current_source in list(self._mapping.keys()):
```

Comment:
A few things here. `dict.keys()` returns a list, so the cast is superfluous. Also, a dictionary supports the in operator so better would be:

    if current_source in self._mapping: